onnxruntime/core/session/inference_session.h

```diff
@@ -628,14 +642,19 @@ class InferenceSession {
   void InitLogger(logging::LoggingManager* logging_manager);
 
   [[nodiscard]] common::Status CheckShapes(const std::string& input_name, const TensorShape& input_shape,
-                                           const TensorShape& expected_shape) const;
+                                           const TensorShape& expected_shape, const char* input_output_moniker) const;
 
   [[nodiscard]] common::Status ValidateInputs(gsl::span<const std::string> feed_names,
                                               gsl::span<const OrtValue> feeds) const;
 
   [[nodiscard]] common::Status ValidateOutputs(gsl::span<const std::string> output_names,
                                                const std::vector<OrtValue>* p_fetches) const;
 
+  [[nodiscard]] common::Status ValidateInputsOutputs(gsl::span<const std::string> feed_fetches_names,
+                                                     gsl::span<const OrtValue> feeds_fetches,
+                                                     const InputOutputDefMetaMap& input_output_meta_map,
+                                                     bool is_inputs) const;
```

Comment:
I like bool in this case, because it is binary. To handle enums properly, one needs to setup a switch with `default` clause and assert there to guard against newly added enums.

Reply:
that's a good point about enums in general. I think there will not be any additional enum values in this case. we can rename it to something like `IOType` or `InputOrOutput` if that makes it more obvious.

on the other hand, `bool` parameters can be ambiguous at call sites, especially if more of them get added later.